### PR TITLE
Remove exceptions-related functionality

### PR DESCRIPTION
**Pull request by ptheof**
Remove support for configuration regarding protected
words and stemming exceptions.
This functionality is provided by the keyword marker
and stemmer override Elasticsearch filters.

### CODE REVIEW

**Comment by ptheof** [comment]
Notes
---------
- Changes in `src/test/resources/stemming_samples.txt` have been manually checked and found sane (i.e. all have to do with protected words and stemming exceptions, and thus were expected). These changes are automatically generated by running `UpdateStemmingSamples.main` method.
- Unit/module tests have been sliced in order to avoid including any protected words and stemming exceptions; this was certainly coarse-grained, maybe some non-exceptional examples have been excluded too. Would you think we should try more delicate cuts here?
- This will not be merged upon PR approval, but according to our deployment plan.
- I am not currently sure about which is the proper plugin version to use; please comment.

File: src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemTokenFilter.java [hunk 1]
```diff
@@ -19,13 +19,12 @@
  * Greek stemmer.
  */
 public class SkroutzGreekStemTokenFilter extends TokenFilter {
-  private final SkroutzGreekStemmer stemmer;
```
Comment (cmantas) [inline src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemTokenFilter.java:22]: Is this change necessary here?

File: src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java [hunk 2]
```diff
@@ -1,19 +1,12 @@
...
 import java.util.Arrays;
-import java.util.List;
 
 import org.apache.lucene.analysis.el.GreekLowerCaseFilter;
-import org.apache.lucene.analysis.CharArrayMap;
 import org.apache.lucene.analysis.CharArraySet;
 
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.common.logging.ESLoggerFactory;
-import org.yaml.snakeyaml.Yaml;
```
Comment (cmantas) [inline src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java:16]: [?] how is this not a dependency found in  `pom.xml` ? :-|

**Review by cmantas — APPROVED** [review]
let's remove the dependency and merge this

File: src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemTokenFilter.java (hunk 1 above)
Comment (ptheof) [reply src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemTokenFilter.java:22]: No. But I don't see why to keep this initialization in the constructor as far as it does not depend on the `input` parameter (nor on `termAtt` and `keywordAttr`). It seems nice to me to give it an equal place in the initializations chain... Truth is that maybe this change should follow the removal of the stemming exceptions functionality on a separate commit; conceptually it does not really belong here :) Would you prefer to revert this?

**Comment by greenonion** [comment]
We should update the README accordingly. Probably make it part of this PR @ptheof

File: src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemTokenFilter.java (hunk 1 above)
Comment (greenonion) [reply src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemTokenFilter.java:22]: I'm afraid I don't have enough context to judge this change, but for what I can tell the behavior is not strictly the same. Prior to the change each `SkroutzGreekStemTokenFilter` instance would initialize its own `SkroutzGreekStemmer` instance. Now they will all share one. I don't know whether we care, I just wanted to point it out :)

By the way, is there something wrong with the indentation? 
You can put it in a separate commit if you want, it's up to you.

File: src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java (hunk 2 above)
Comment (ptheof) [reply src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java:16]: Good question. Because it comes from `org.elasticsearch`:
```
[INFO] gr.skroutz:elasticsearch-skroutz-greekstemmer:jar:5.4.2.6                                         
[INFO] +- org.elasticsearch:elasticsearch:jar:5.4.2:compile                            
[INFO] |  +- org.apache.lucene:lucene-core:jar:6.5.1:compile                                             
[INFO] |  +- org.apache.lucene:lucene-analyzers-common:jar:6.5.1:compile                                 
[INFO] |  +- org.apache.lucene:lucene-backward-codecs:jar:6.5.1:compile                                  
[INFO] |  +- org.apache.lucene:lucene-grouping:jar:6.5.1:compile                                         
[INFO] |  +- org.apache.lucene:lucene-highlighter:jar:6.5.1:compile                                      
[INFO] |  +- org.apache.lucene:lucene-join:jar:6.5.1:compile                                             
[INFO] |  +- org.apache.lucene:lucene-memory:jar:6.5.1:compile                                           
[INFO] |  +- org.apache.lucene:lucene-misc:jar:6.5.1:compile                                             
[INFO] |  +- org.apache.lucene:lucene-queries:jar:6.5.1:compile                                          
[INFO] |  +- org.apache.lucene:lucene-queryparser:jar:6.5.1:compile                                      
[INFO] |  +- org.apache.lucene:lucene-sandbox:jar:6.5.1:compile
[INFO] |  +- org.apache.lucene:lucene-spatial:jar:6.5.1:compile                                          
[INFO] |  +- org.apache.lucene:lucene-spatial-extras:jar:6.5.1:compile
[INFO] |  +- org.apache.lucene:lucene-spatial3d:jar:6.5.1:compile                                        
[INFO] |  +- org.apache.lucene:lucene-suggest:jar:6.5.1:compile
[INFO] |  +- org.elasticsearch:securesm:jar:1.1:compile                                                  
[INFO] |  +- net.sf.jopt-simple:jopt-simple:jar:5.0.2:compile                                            
[INFO] |  +- com.carrotsearch:hppc:jar:0.7.1:compile                                                     
[INFO] |  +- joda-time:joda-time:jar:2.9.5:compile                                                       
[INFO] |  +- org.yaml:snakeyaml:jar:1.15:compile
...
```
As the package had been used in `SkroutzGreekStemmer` directly, the dependency should be explicitly declared. In any case, now there is no such dependency at all, and `pom.xml` remains as is... :)

**Review by greenonion — CHANGES_REQUESTED** [review]
LGTM overall, nice job :)

Let's decide on the constructor thing, and please add a README entry on how to define stemming exceptions. We could (should?) also maybe include a sample exceptions file in the repo.

File: src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemTokenFilter.java (hunk 1 above)
Comment (ptheof) [reply src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemTokenFilter.java:22]: Hmm... No, in Java you need the `static` modifier to have a "class member" (i.e. common for all objects of this class). As the constructor is not a method in Java, you have to use `this` there in any case. By taking the initialization of the variable out of the constructor nothing changes regarding the kind of the variable. The only thing it does change here is the order of the initializations. Inline initializations go first, followed by the code in the constructor (in fact, we can imagine it as if all inline initialization are put in order at the top of the constructor body). So, in our case we have the initializations taking place in this order:
`stemmer`, `termAtt`, `keywordAttr`, <whatever else `super` initializes`,
while previously we had:
`termAtt`, `keywordAttr`, <whatever else `super` initializes`, `stemmer`
In the absence of dependencies between the initializations, all these are fine from a functional POV.

Anyway, I can certainly keep back this change (or maybe put it on another commit) if it creates any confusion---it was just a spontaneous reaction when my eye fell on this piece of code!

Comment (greenonion) [reply src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemTokenFilter.java:22]: Very nice, thanks for clearing it up for me :) Let's go forward with this change, it makes sense.

**Comment by ptheof** [comment]
Ok @greenonion, added a small section in README, and also included a sample configuration file for the stemmer-override ES plugin.

File: src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java [hunk 3]
```diff
@@ -951,13 +953,22 @@ private int rule22(char s[], int len) {
...
+        return len - 5;
       } else {
         return len - 2;
       }
     }
 
-    if (len > 1 && endsWithVowel(s, len))
-      return len - 1;
+    if (String.valueOf(s).equals("πασχα"))
+      return len;
```
Comment (cmantas) [inline src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java:964]: why not return 5?
actually why not 
```java
if (String.valueOf(s).startsWith("πασχα) return 5
```
somewhere before  L386 ?

Comment (ptheof) [reply src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java:964]: Generally you cannot short-circuit the stemmer this way. Observe that all rules except for `#22` are applied unconditionally, and therefore handling "πασχα*" --> "πασχα" in one rule does not ensure e.g. that "πασχα" will not be further stemmed to "πασχ". In order to preserve the current design of the stemming algorithm, I thought it was better to find the specific rules that currently affect the stemming of the words we are interested in, and add some handling for these words there.

File: src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java [hunk 4]
```diff
@@ -951,13 +953,22 @@ private int rule22(char s[], int len) {
       } else if(String.valueOf(s).startsWith("ντουλαπ")) {
         s[len - 2] = 'α';
         return len - 1;
+      } else if (String.valueOf(s).startsWith("πασχαλιν")) {
+        return len - 5;
```
Comment (greenonion) [inline src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java]: Will this also cover "πασχαλιατικη"/"πασχαλιατικο";

File: src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java [hunk 5]
```diff
@@ -388,6 +388,8 @@ private int rule6(char s[], int len) {
           exc6.contains(s, 0, len) ||
           endsWith(s, len, "φοιν"))
         len += 2; // add back -ικ
+      else if (String.valueOf(s).startsWith("πασχαλιατ"))
```
Comment (greenonion) [inline src/main/java/org/elasticsearch/index/analysis/SkroutzGreekStemmer.java]: Why is this needed, I don't get it! Aren't all cases covered by the first or second branches of the previous conditional? (I'm obviously missing something here.)

**Review by greenonion — APPROVED** [review]
Very nice, LGTM!
